### PR TITLE
SS22 local competition

### PR DESCRIPTION
**Pull request by kvnptl**
## Changelog

* [Add] perception - YOLOv7 for rgb_object_recognition in PyTorch
* [Add] perception - empty space detection
* [Add] manipulation - default, side grasp, top grasp, orientation independent IK
* [Add] manipulation - grasp verification from gripper motors
* [Add] robot configuration - support for tower camera and arm camera configuration
* [Enhancements] ROS package:
  - mir_object_recognition
  - mir_actions - support for yb2 and yb4 robot configuration, etc.
  - mir_planning
 
## Checklist:

- [x] My code doesn't contain unnecessary comment blocks (e.g. unused code, templates of `package.xml` or `CMakeLists.txt`)
- [x] I have updated the `package.xml` and `CMakeLists.txt` with the correct dependencies.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: mir_manipulation/mir_grasp_monitors/ros/scripts/youbot_dynamixel_gripper_grasp_monitor_node [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        while self.received_state == False:
+            rospy.sleep(0.01)
+
+        return
+
+    def is_gripper_closed(self, req):
+        self.wait_for_new_joint_states()
+
+        for gripper_joint_value in self.current_joint_configuration:
+            if gripper_joint_value > 0.003:
```
Comment (deebuls) [inline mir_manipulation/mir_grasp_monitors/ros/scripts/youbot_dynamixel_gripper_grasp_monitor_node:58]: What  is this 0.003  ? 
Make it as a config parameter better a dynamic configurable parameter.

File: mir_manipulation/mir_pregrasp_planning/ros/src/mir_pregrasp_planning_ros/pregrasp_planner_node.py [hunk 2]
```diff
@@ -276,6 +281,103 @@ def idle_state(self):
...
+            return False
+
+    def orientation_independent_ik_routine(self, transformed_pose, grasp_type):
+        rospy.loginfo("[Pregrasp Planning] Tryinng orientation independent planning.")
+        input_pose = geometry_msgs.msg.PoseStamped()
+        input_pose.header = transformed_pose.header
+        input_pose.pose.position = transformed_pose.pose.position
+        if grasp_type == "side_grasp":
+            # input_pose.pose.position.y -= 0.01
+            input_pose.pose.position.x -= 0.01
```
Comment (deebuls) [inline mir_manipulation/mir_pregrasp_planning/ros/src/mir_pregrasp_planning_ros/pregrasp_planner_node.py]: Add a comment on hwy this is being done

File: mir_manipulation/mir_pregrasp_planning/ros/src/mir_pregrasp_planning_ros/pregrasp_planner_node.py [hunk 3]
```diff
@@ -346,50 +426,52 @@ def running_state(self):
         else:
             grasp_type = "side_grasp"
 
+        if grasp_type == "side_grasp":
```
Comment (deebuls) [inline mir_manipulation/mir_pregrasp_planning/ros/src/mir_pregrasp_planning_ros/pregrasp_planner_node.py]: Comments on why you are adding these values and what is expected behaviour.

File: mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py [hunk 4]
```diff
@@ -59,26 +52,48 @@ def classify(self, pointcloud, center=True, rotate=True, pad=True):
...
-        feed_dict = {self.ops['pointcloud']: pointcloud,
-                     self.ops['is_training']: False}
-        
-        pred_label, probs = self.sess.run([self.ops['predictions'], self.ops['probabilities']],
-                                         feed_dict=feed_dict)
-        probs = probs[0][pred_label][0]
-
-        return pred_label[0], probs 
\ No newline at end of file
+        # Kevin: this code is returning len 1000 instead of len 1?
```
Comment (deebuls) [inline mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py]: Delete this

File: mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node [hunk 5]
```diff
@@ -79,22 +88,43 @@ class PointcloudObjectRecognizer():
...
                     name, probability = self.classifier.classify(features)
                     name = name[0]
                 elif self.model == "cnn_based":
-                    cloud = self.extract_pointcloud(object.views[0].point_cloud, color="rgb")
-                    label, probability = self.classifier.classify(cloud, center=True, rotate=True, pad=True)
-                    name = self.label_to_name[label]
+                    cloud = self.extract_pointcloud(
+                        object.views[0].point_cloud, color="rgb")
+
+                    # Kevin: something wrong with passing the cloud to the classifier
```
Comment (deebuls) [inline mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node]: check the comments

**Review by deebuls — CHANGES_REQUESTED** [review]

File: mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py (hunk 4 above)
Comment (RavisankarSelvaraju) [reply mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py]: deleted

File: mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node (hunk 5 above)
Comment (RavisankarSelvaraju) [reply mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node]: comments removed

File: mir_manipulation/mir_pregrasp_planning/ros/src/mir_pregrasp_planning_ros/pregrasp_planner_node.py (hunk 3 above)
Comment (RavisankarSelvaraju) [reply mir_manipulation/mir_pregrasp_planning/ros/src/mir_pregrasp_planning_ros/pregrasp_planner_node.py]: - values are modified as dynamics config params
- added comment for the values

File: mir_manipulation/mir_grasp_monitors/ros/scripts/youbot_dynamixel_gripper_grasp_monitor_node (hunk 1 above)
Comment (RavisankarSelvaraju) [reply mir_manipulation/mir_grasp_monitors/ros/scripts/youbot_dynamixel_gripper_grasp_monitor_node:58]: We are not using this node for gripper
- This package is for old robot configuration and currently not in use

File: mir_manipulation/mir_moveit_client/ros/launch/moveit_client.launch [hunk 6]
```diff
@@ -3,7 +3,7 @@
     <node pkg="mcr_moveit_client" type="moveit_client_node" name="arm_moveit_client" output="screen">
         <rosparam command="load" file="$(find mir_moveit_client)/ros/config/arm_params.yaml"/>
     </node>
-    <node pkg="mcr_moveit_client" type="moveit_client_node" name="gripper_moveit_client" output="screen">
+    <!-- <node pkg="mcr_moveit_client" type="moveit_client_node" name="gripper_moveit_client" output="screen">
```
Comment (sthoduka) [inline mir_manipulation/mir_moveit_client/ros/launch/moveit_client.launch]: remove commented lines

File: mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py [hunk 7]
```diff
@@ -59,26 +52,46 @@ def classify(self, pointcloud, center=True, rotate=True, pad=True):
...
+            self.model_path, map_location=torch.device("cuda" if torch.cuda.is_available() else "cpu")))
+        model = model.eval()
+
+        break_flag = False
+
+        for pcl, label in pointcloud_dataloader:
+
+            if break_flag == False:
+
+                print("inferencing ...")
```
Comment (sthoduka) [inline mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py]: remove blank lines

File: mir_perception/mir_object_recognition/common/src/rgb_object_recognition/__init__.py [hunk 8]
```diff
@@ -1,3 +1,5 @@
-from squeezedet_classifier import SqueezeDetClassifier
-from ssdlite_mobilenet import SSDLiteMobilenet
-from utils import util
\ No newline at end of file
+# from .squeezedet_classifier import SqueezeDetClassifier
+# from .ssdlite_mobilenet import SSDLiteMobilenet
+# from .utils import util
+# from image_classifier import ImageClassifier
+# from .yolov7 import yolov7_classifier
```
Comment (sthoduka) [inline mir_perception/mir_object_recognition/common/src/rgb_object_recognition/__init__.py]: is the last import needed or not? If the others are not used, you can delete them

File: mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node [hunk 9]
```diff
@@ -79,22 +88,39 @@ class PointcloudObjectRecognizer():
...
+                    if (label, probability) != (None, None):
+                        name = self.label_to_name[label]
+                        probability = probability
+                    else:
+                        name = "unknown object found"
+                        probability = 0.0
+                        continue
+
+                    print('\n######################')
+                    print('name: ', name)
```
Comment (sthoduka) [inline mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node]: is this needed? If yes, replace by rospy.loginfo or logdebug

File: mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node [hunk 10]
```diff
@@ -1,114 +1,183 @@
...
+                            result.probability = probs[i]
+                            roi = RegionOfInterest()
+                            roi.x_offset = int(bboxes[i][0])
+                            roi.y_offset = int(bboxes[i][1])
+                            roi.width = int(bboxes[i][2]) - int(bboxes[i][0])
+                            roi.height = int(bboxes[i][3]) - int(bboxes[i][1])
+                            result.roi = roi
+                            objects.append(result)
+                        if self.debug:
+                            print("DETECTED OBJECTS: ", labels)
```
Comment (sthoduka) [inline mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node]: rospy.logdebug or loginfo

File: mir_perception/mir_perception_utils/ros/src/object_utils_ros.cpp [hunk 11]
```diff
@@ -70,11 +70,33 @@ void object::estimatePose(const PointCloud::Ptr &xyz_input_cloud, geometry_msgs:
...
+
+  ////////////////////
+  // publish filtered cloud on topic for debugging
+  ///////////////////
+
+  // #include <sensor_msgs/PointCloud2.h>
+
+  // ros::NodeHandle nh("~");
+  // ros::Publisher pub_filtered_cloud_;
+  // pub_filtered_cloud_ = nh.advertise<sensor_msgs::PointCloud2>("CUSTOM_pub_filtered_cloud_", 1);
```
Comment (sthoduka) [inline mir_perception/mir_perception_utils/ros/src/object_utils_ros.cpp]: remove commented code (or if it's necessary, add it under a debug condition)

**Comment by sthoduka** [comment]
Mostly looks good to me. Some minor changes mentioned in the comments. I think the object recognition package needs to be cleaned up, but can be done later (e.g. with dcgnn versions).

File: mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node (hunk 10 above)
Comment (RavisankarSelvaraju) [reply mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node]: done

File: mir_perception/mir_perception_utils/ros/src/object_utils_ros.cpp (hunk 11 above)
Comment (RavisankarSelvaraju) [reply mir_perception/mir_perception_utils/ros/src/object_utils_ros.cpp]: Removed

File: mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node (hunk 9 above)
Comment (RavisankarSelvaraju) [reply mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node]: replaced with rospy.loginfo

File: mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py (hunk 7 above)
Comment (RavisankarSelvaraju) [reply mir_perception/mir_object_recognition/common/src/pc_object_recognition/dgcnn_classifier.py]: blank lines Removed

File: mir_manipulation/mir_moveit_client/ros/launch/moveit_client.launch (hunk 6 above)
Comment (RavisankarSelvaraju) [reply mir_manipulation/mir_moveit_client/ros/launch/moveit_client.launch]: Done

File: mir_perception/mir_object_recognition/common/src/rgb_object_recognition/__init__.py (hunk 8 above)
Comment (RavisankarSelvaraju) [reply mir_perception/mir_object_recognition/common/src/rgb_object_recognition/__init__.py]: Removed

File: mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node [hunk 12]
```diff
@@ -149,7 +149,7 @@ class RGBObjectRecognizer():
                             result.roi = roi
                             objects.append(result)
                         if self.debug:
-                            print("DETECTED OBJECTS: ", labels)
+                            rospy.logdebug("DETECTED OBJECTS: ", labels)
```
Comment (sthoduka) [inline mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node]: You need the string formatting operator: `rospy.logdebug('Detected objects: %s', labels)`

File: mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node [hunk 13]
```diff
@@ -116,11 +113,7 @@ class PointcloudObjectRecognizer():
                         name = "unknown object found"
                         probability = 0.0
                         continue
-
-                    print('\n######################')
-                    print('name: ', name)
-                    print('######################\n')
-
+                    rospy.loginfo('Name of the object found: ',name)
```
Comment (sthoduka) [inline mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node]: same here: `rospy.loginfo('Name of the object found: %s',name)`

File: mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node (hunk 12 above)
Comment (vamsikalagaturu) [reply mir_perception/mir_object_recognition/ros/scripts/rgb_object_recognizer_node]: resolved

File: mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node (hunk 13 above)
Comment (vamsikalagaturu) [reply mir_perception/mir_object_recognition/ros/scripts/pc_object_recognizer_node]: resolved

**Comment by sthoduka** [comment]
thanks all
